### PR TITLE
Adding CredentialOptions + moving Credential retrieval to BigtableSession

### PR DESCRIPTION
**Pull request by sduskis**

### CODE REVIEW

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+package com.google.cloud.bigtable.config;
+
+import com.google.common.base.Preconditions;
+
+public class CredentialOptions {
+  public static final String SERVICE_ACCOUNT_JSON_ENV_VARIABLE = "GOOGLE_APPLICATION_CREDENTIALS";
+  protected static final Logger LOG = new Logger(CredentialOptions.class);
+
+  public enum CredentialType {
+    Json, MetadataServiceAccount, None, P12,
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java]: can you explain/document each type?

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+  public static boolean hasJsonCredentialConfigured() {
+    return System.getenv().containsKey(SERVICE_ACCOUNT_JSON_ENV_VARIABLE);
+  }
+
+  public static CredentialOptions metadataServiceAccountCredential() {
+    return new CredentialOptions(CredentialType.MetadataServiceAccount);
+  }
+
+  public static CredentialOptions noCredential() {
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java]: nit: can it be a little more consistent here: "no", "null" or "none".

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  private final String serviceAccount;
+  private final String keyFile;
+
+  private CredentialOptions(CredentialType credentialType) {
+    this.credentialType = credentialType;
+    this.serviceAccount = null;
+    this.keyFile = null;
+  }
+
+  private CredentialOptions(String serviceAccount, String keyFile) {
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java]: Can we have following:
private CredentialOptions(CredentialType credentialType) {
    CredentialOptions(credentialType, null, null);
}

private CredentialOptions(credentialType, serviceAccount, keyFile) {
   ...
}

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    this.serviceAccount = serviceAccount;
+    this.keyFile = keyFile;
+  }
+
+  public CredentialType getCredentialType() {
+    return credentialType;
+  }
+
+  /**
+   * ServiceAccount email address used for the P12 file -
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java]: ServiceAccount is used in json authentication as well, it is within JSON key file.

ServiceAccount email address used along with the P12 file?

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+   * https://developers.google.com/identity/protocols/OAuth2ServiceAccount
+   */
+  public String getServiceAccount() {
+    Preconditions.checkArgument(credentialType == CredentialType.P12,
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java]: don't check it in get() method, do condition check when we setting it.

this.credentialType = credentialType;
this.serviceAccount = serviceAccount;
this.keyFile = keyFile;

if (credentialType == CredentialType.P12) {
  Preconditions.checkArgument(serviceAccount != null && keyFile != nul);
} else {
  Preconditions.checkArgument(serviceAccount == null && keyFile == nul);
}

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  public String getKeyFile() {
+    Preconditions.checkArgument(credentialType == CredentialType.P12,
+        "Only P12 credentials have a key file");
+
+    return keyFile;
+  }
+
+  public String getJsonFile() {
+    Preconditions.checkArgument(credentialType == CredentialType.Json,
+        "Only Json credentials have a key file");
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java]: "Only P12 credentials have a key file" and "Only Json credentials have a key file"?

When we set CredentialType:
if (credentialType == CredentialType.P12) {
   Preconditions.checkArgument(hasJsonCredentialConfigured());
}

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java [hunk 7]
```diff
@@ -220,20 +226,45 @@ protected static ScheduledExecutorService createDefaultRetryExecutor() {
...
   public BigtableSession(BigtableOptions options, ExecutorService batchPool) {
-    this(options, batchPool, createDefaultEventLoopGroup(), createDefaultRetryExecutor());
+    this(options, batchPool, null, null);
   }
 
-  public BigtableSession(BigtableOptions options, ExecutorService batchPool, EventLoopGroup elg,
-      ScheduledExecutorService scheduledRetries) {
-    this.elg = elg;
+  public BigtableSession(final BigtableOptions options, final ExecutorService batchPool,
+      EventLoopGroup elg, ScheduledExecutorService scheduledRetries) {
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: nit: @nullable EventLoopGroup and ScheduledExecutorService

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java [hunk 8]
```diff
@@ -220,20 +226,45 @@ protected static ScheduledExecutorService createDefaultRetryExecutor() {
...
+              }
+            }
+          });
+          return interceptor;
+        } else {
+          return null;
+        }
+      }
+    });
+    this.elg = elg == null ? createDefaultEventLoopGroup() : elg;
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: nit: (elg == null)

**Comment by AngusDavis** [comment]
A couple of requests: 1) A general CredentialOptions#withCredentials(Credentials) or some such for BYOCreds 2) A way to do JSON credentials not from environment variables (e.g, passing a file path that the file should be read from).

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java [hunk 9]
```diff
@@ -220,20 +226,45 @@ protected static ScheduledExecutorService createDefaultRetryExecutor() {
...
+              try {
+                credentials.refresh();
+              } catch (IOException e) {
+                LOG.error("Could not initialize the credentials", e);
+              }
+            }
+          });
+          return interceptor;
+        } else {
+          return null;
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: Can credentials be null? Can you document or log warning?

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java [hunk 10]
```diff
@@ -220,20 +226,45 @@ protected static ScheduledExecutorService createDefaultRetryExecutor() {
...
+      }
+    });
+    this.elg = elg == null ? createDefaultEventLoopGroup() : elg;
     this.batchPool = batchPool;
     this.options = options;
-    LOG.info("Opening connection for project %s on data host %s, "
-        + "table admin host %s.", options.getProjectId(),
-      options.getDataHost(), options.getTableAdminHost());
-    this.scheduledRetries = scheduledRetries;
+    LOG.info("Opening connection for project %s on data host %s, " + "table admin host %s.",
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: u don't need " + "

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java (hunk 6 above)
Comment (sduskis) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java]: The setter seems to be fine.  I'll remove the preconditions on the getters.

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java (hunk 9 above)
Comment (sduskis) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/grpc/BigtableSession.java]: The only time they can be null is for unit testing.  See CredentialFactory.nullCredential() for the warning.

**Comment by sduskis** [comment]
I updated the logic.  PTAL.

File: bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    public String getKeyFile() {
+      return keyFile;
+    }
+  }
+
+  /**
+   * A CredentialOption that supplies the Credentials directly.
+   */
+  public static class UserSuppliedCredentialOptions extends CredentialOptions {
+    private Credentials credential;
```
Comment (kevinsi4508) [inline bigtable-grpc-interface/src/main/java/com/google/cloud/bigtable/config/CredentialOptions.java]: nit: private final

**Comment by kevinsi4508** [comment]
LGTM

**Pull request merged**